### ISSUE TITLE
LNK1181 error building bundled libzmq on MSVC 14.29 due to incorrect LIBPATH

### ISSUE THREAD

**Issue by gtowers-dukosi** [zeromq/pyzmq#1694]
`pip install --no-binary pyzmq pyzmq` fails with a LNK1181 error when attempting to build the libzmq library with environmental variable `ZMQ_PREFIX=bundled` on Windows 10 21H2 using Python 3.10.4 with pip 22.0.4, setuptools 62.1.0 and wheel 0.37.1, building using MSVC 14.29.30133.

The error was originally observed when attempting  `pip install pyzmq==22.0.1 --install-option="--zmq=bundled"` on python 3.9, but  was then replicated with the latest PyZMQ (v22.3.0) using Python 3.10 and the latest pip, setuptools and wheel as directed on https://github.com/zeromq/pyzmq/wiki/Building-and-Installing-PyZMQ

The wheel fails to build with the following error: `LINK : fatal error LNK1181: cannot open input file 'libzmq.cp310-win_amd64.lib'`

This appears to be caused by an error in the library path provided to the MSVC linker: `/LIBPATH:build\temp.win-amd64-3.10\Release\buildutils` is provided to the linker, but the relevant .lib file is stored in `build\temp.win-amd64-cpython-310\Release\buildutils`

Console output:

```
    Creating library build\temp.win-amd64-cpython-310\Release\buildutils\libzmq.cp310-win_amd64.lib and object build\temp.win-amd64-cpython-310\Release\buildutils\libzmq.cp310-win_amd64.exp
      Generating code
      Finished generating code
      building 'zmq.backend.cython._device' extension
      creating build\temp.win-amd64-cpython-310\Release\zmq
      creating build\temp.win-amd64-cpython-310\Release\zmq\backend
      creating build\temp.win-amd64-cpython-310\Release\zmq\backend\cython
      "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DWINVER=0x0602 -D_WIN32_WINNT=0x0602 -Ibundled\zeromq\include -Izmq\utils -IC:\Python310\include -IC:\Python310\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" /Tczmq\backend\cython\_device.c /Fobuild\temp.win-amd64-cpython-310\Release\zmq\backend\cython\_device.obj
      _device.c
      "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\link.exe" /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:build\temp.win-amd64-3.10\Release\buildutils /LIBPATH:C:\Python310\libs /LIBPATH:C:\Python310 /LIBPATH:C:\Python310\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.19041.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.19041.0\um\x64" libzmq.cp310-win_amd64.lib ws2_32.lib iphlpapi.lib advapi32.lib /EXPORT:PyInit__device build\temp.win-amd64-cpython-310\Release\zmq\backend\cython\_device.obj /OUT:build\lib.win-amd64-cpython-310\zmq\backend\cython\_device.cp310-win_amd64.pyd /IMPLIB:build\temp.win-amd64-cpython-310\Release\zmq\backend\cython\_device.cp310-win_amd64.lib
      LINK : fatal error LNK1181: cannot open input file 'libzmq.cp310-win_amd64.lib'
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Professional\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\link.exe' failed with exit code 1181
```

I can't see why the build directory name is being corrupted before being passed into the MSVC linker. Is this an issue with the wheel/setup of pyzmq or could this be related to the version of MSVC/Windows Kit used?

Full console log also attached: 
[Pyzmq_Python310_LNK1181_Error.txt](https://github.com/zeromq/pyzmq/files/8620767/Pyzmq_Python310_LNK1181_Error.txt)

**Comment by minrk** [comment]
I think this may have been fixed by bec362447c0bd5002cf813894a2fec3e8ee8ce5d.

Can you test with `pip install --pre pyzmq` to see if the pyzmq-23 beta works for you?

**Comment by gtowers-dukosi** [comment]
Yeah, I think you're right that it was fixed in bec3624.
The pyzmq-23 beta works correctly. Looks like newer versions of setuptools have a different format of build_temp and it didn't match the string assembly used prior to the bec3624 patch.

**Comment by minrk** [comment]
okay, thanks. Closing, then. Planning to push out 23 pretty soon (hoping for Cython 0.29.29 first, to get Python 3.11-compatible sources)
